torch/csrc/autograd/saved_variable.h

```diff
@@ -1,46 +1,68 @@
 #pragma once
 
-#include <mutex>
-#include <memory>
-#include <functional>
-#include <ATen/ATen.h>
-
-#include "torch/csrc/jit/tracer_state.h"
-#include "torch/csrc/autograd/variable.h"
 #include "torch/csrc/autograd/variable_version.h"
-#include "torch/csrc/Types.h"
 
-namespace torch { namespace autograd {
+#include <ATen/Tensor.h>
+
+#include <cstdint>
+#include <list>
+#include <memory>
 
+namespace torch {
+namespace autograd {
+class Variable;
 struct Function;
+} // namespace autograd
+namespace jit { namespace tracer {
+struct ValueTracingStateElem;
+using ValueTracingState = std::list<ValueTracingStateElem>;
+}} // namespace jit::tracer
+} // namespace torch
 
-extern const char* ERR_BACKWARD_TWICE;
+namespace torch { namespace autograd {
 
-struct SavedVariable {
-  SavedVariable()
-    : data()
-    , has_grad_fn(false)
-    , version()
-    , requires_grad(false)
-    , expected_version(-1) {}
+extern const char* ERR_BACKWARD_TWICE;
 
+/// A snapshot of a variable at a certain version. A `SavedVariable` stores
+/// enough information to reconstruct a variable from a certain point in time.
+class SavedVariable {
+ public:
+  SavedVariable() = default;
   SavedVariable(const Variable& variable, bool is_output);
+  SavedVariable(SavedVariable&&) = default;
+  SavedVariable& operator=(SavedVariable&&) = default;
+
+  // Must be defined externally to avoid it being inlined by the compiler,
+  // which would require it to see the definition of ValueTracingState.
+  ~SavedVariable();
```

Comment:
Why? It was automatically defined previously and worked fine

Reply:
I forward declared `ValueTracingStateElem` so we don't have to include `tracer_state.h` which includes the whole JIT IR. Holding a `shared_ptr`/`unique_ptr` to an incomplete type works only if the destructor is defined in the implementation file (where the definition of `ValueTracingStateElem` is available), as the inlined destructor will otherwise call the destructor of the smart pointer, which will require the size/definition of the forward declared class.

Reply:
Does this mean that Variable.h will no longer include ir.h? That is a big win for compilation times.

Reply:
This feels like tail wagging the dog. Wouldn't you rather have a complete definition for `ValueTracingStateElem` and incomplete types for `TracingState` and `Value`? (does the destructor problem apply to `weak_ptr` too?)

Reply:
Yeah I'd rather have `ValueTracingStateElem` to be visible, with a forward declaration of required JIT types then have to change the Variable.